pointmatcher/DataPointsFilters/Geometry.cpp

```diff
@@ -0,0 +1,170 @@
+// vim: ts=4:sw=4:noexpandtab
+/*
+
+Copyright (c) 2010--2018,
+François Pomerleau and Stephane Magnenat, ASL, ETHZ, Switzerland
+You can contact the authors at <f dot pomerleau at gmail dot com> and
+<stephane at magnenat dot net>
+
+All rights reserved.
+
+Redistribution and use in source and binary forms, with or without
+modification, are permitted provided that the following conditions are met:
+    * Redistributions of source code must retain the above copyright
+      notice, this list of conditions and the following disclaimer.
+    * Redistributions in binary form must reproduce the above copyright
+      notice, this list of conditions and the following disclaimer in the
+      documentation and/or other materials provided with the distribution.
+    * Neither the name of the <organization> nor the
+      names of its contributors may be used to endorse or promote products
+      derived from this software without specific prior written permission.
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+DISCLAIMED. IN NO EVENT SHALL ETH-ASL BE LIABLE FOR ANY
+DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+(INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+*/
+#include "Geometry.h"
+
+// Eigenvalues
+#include "Eigen/QR"
+#include "Eigen/Eigenvalues"
+
+#include "PointMatcherPrivate.h"
+#include "IO.h"
+#include "MatchersImpl.h"
+
+#include <boost/format.hpp>
+
+
+#include "utils.h"
+
+// SurfaceNormalDataPointsFilter
+// Constructor
+template<typename T>
+GeometryDataPointsFilter<T>::GeometryDataPointsFilter(const Parameters& params):
+	PointMatcher<T>::DataPointsFilter("GeometryDataPointsFilter",
+	        GeometryDataPointsFilter::availableParameters(), params),
+    keepUnstructureness(Parametrizable::get<int>("keepUnstructureness")),
+    keepStructureness(Parametrizable::get<T>("keepStructureness"))
+{
+}
+
+// Compute
+template<typename T>
+typename PointMatcher<T>::DataPoints
+GeometryDataPointsFilter<T>::filter(
+	const DataPoints& input)
+{
+	DataPoints output(input);
+	inPlaceFilter(output);
+	return output;
+}
+
+// In-place filter
+template<typename T>
+void GeometryDataPointsFilter<T>::inPlaceFilter(
+	DataPoints& cloud)
+{
+	typedef typename DataPoints::View View;
+	typedef typename DataPoints::Label Label;
+	typedef typename DataPoints::Labels Labels;
+
+	const int pointsCount(cloud.features.cols());
+	const int descDim(cloud.descriptors.rows());
```

Comment:
Why use signed int instead of an unsigned int? You leave half of the bits unused.

I would suggest using size_t: https://www.geeksforgeeks.org/size_t-data-type-c-language/

Reply:
Yep, good point.